tcpdialer.go

```diff
@@ -327,7 +327,11 @@ func (d *TCPDialer) tryDial(network string, addr *net.TCPAddr, deadline time.Tim
 		defer func() { <-concurrencyCh }()
 	}
 
-	dialer := net.Dialer{LocalAddr: d.LocalAddr}
+	dialer := net.Dialer{}
+	if d.LocalAddr != nil {
+		dialer = net.Dialer{LocalAddr: d.LocalAddr}
+	}
```

Comment:
I'm sorry but this doesn't change anything? Fields are `nil` by default so with `net.Dialer{}` `LocalAddr` will be `nil` but when `d.LocalAddr` is `nil` and you do `net.Dialer{LocalAddr: d.LocalAddr}` it will be exactly the same as doing `net.Dialer{}`.

Reply:
Actually, they are not the name. This is because `net.Dialer.LocalAddr` is an interface type while the `d.LocalAddr` is a concrete type.  The statement `net.Dialer{LocalAddr: d.LocalAddr}` with  nil`d.LocalAddr`  will set the underlying type in `Dialer.LocalAddr` to `*net.TCPAddr` and the data to nil,  which lead nil check for `Dialer.LocalAddr` to fail.

There is a test sample on golang playground - https://play.golang.org/p/E1tlCOb32P